File: Java_Gemini_Flash/Chain_Responsibility/Round1/Code.java
package Chain_of_responsibility.round1;

import java.util.Optional;

interface PackageHandler {
    Optional<String> handle(PackageRequest request);

    PackageHandler nextHandler();

    void setNextHandler(PackageHandler handler);
}

class PackageRequest {
    private String location;
    private double amount;

    public PackageRequest(String location, double amount) {
        this.location = location;
        this.amount = amount;
    }

    public String getLocation() {
        return location;
    }

    public double getAmount() {
        return amount;
    }
}

class LocationHandler implements PackageHandler {
    private PackageHandler nextHandler;

    @Override
    public Optional<String> handle(PackageRequest request) {
        if (request.getLocation().equals("Invalid Location")) {
            return Optional.empty();
        } else {
            return Optional.of("Location Verified");
        }
    }

    @Override
    public PackageHandler nextHandler() {
        return nextHandler;
    }

    @Override
    public void setNextHandler(PackageHandler handler) {
        this.nextHandler = handler;
    }
}

class PaymentHandler implements PackageHandler {
    private PackageHandler nextHandler;

    @Override
    public Optional<String> handle(PackageRequest request) {
        if (request.getAmount() <= 0) {
            return Optional.empty();
        } else {
            return Optional.of("Payment Processed");
        }
    }

    @Override
    public PackageHandler nextHandler() {
        return nextHandler;
    }

    @Override
    public void setNextHandler(PackageHandler handler) {
        this.nextHandler = handler;
    }
}

class DeliveryHandler implements PackageHandler {
    private PackageHandler nextHandler;

    @Override
    public Optional<String> handle(PackageRequest request) {
        return Optional.of("Package Delivered");
    }

    @Override
    public PackageHandler nextHandler() {
        return nextHandler;
    }

    @Override
    public void setNextHandler(PackageHandler handler) {
        this.nextHandler = handler;
    }
}

class PackageDeliveryChain {
    private PackageHandler firstHandler;

    public PackageDeliveryChain(PackageHandler firstHandler) {
        this.firstHandler = firstHandler;
    }

    public String processRequest(PackageRequest request) {
        PackageHandler currentHandler = firstHandler;
        Optional<String> result = Optional.empty();

        while (result.isEmpty() && currentHandler != null) {
            result = currentHandler.handle(request);
            currentHandler = currentHandler.nextHandler();
        }

        if (result.isPresent()) {
            return result.get();
        } else {
            return "Request failed at all handlers.";
        }
    }

    public void setNextHandler(PackageHandler handler) {
        PackageHandler currentHandler = firstHandler;
        while (currentHandler.nextHandler() != null) {
            currentHandler = currentHandler.nextHandler();
        }
        currentHandler.setNextHandler(handler);
    }
}

// Removed AbstractPackageHandler class as it is no longer needed

File: ChatGPT3_5/ChainResponsibility/round1/Code.py
class Handler:
    def __init__(self):
        self.next_handler = None

    def set_next(self, handler):
        self.next_handler = handler
        return handler

    def handle(self, request):
        if self.next_handler:
            return self.next_handler.handle(request)
        return None

class LocationHandler(Handler):
    def handle(self, request):
        if request.get('location') == 'valid_location':
            return "Location verified"
        return super().handle(request)

class PaymentHandler(Handler):
    def handle(self, request):
        if request.get('payment') == 'paid':
            return "Payment processed"
        return super().handle(request)

class DeliveryHandler(Handler):
    def handle(self, request):
        if request.get('delivery_ready') == True:
            return "Package delivered"
        return super().handle(request)

# Example usage
if __name__ == "__main__":
    location_handler = LocationHandler()
    payment_handler = PaymentHandler()
    delivery_handler = DeliveryHandler()

    location_handler.set_next(payment_handler).set_next(delivery_handler)

    request = {'location': 'valid_location', 'payment': 'paid', 'delivery_ready': True}
    result = location_handler.handle(request)
    print(result)  # Output should be "Location verified"
